### PR TITLE
Optimize performance of SIMD binary operations via polymorphic builtins

### PR DESCRIPTION
**Pull request by nvzqz**
These changes allow for expressing the vector semantics of our `SIMD{n}<T>` types directly to LLVM. This is done via polymorphic builtins defined by @gottesmm. These builtins are only called for stdlib types on the condition of `Swift._isConcrete` (calls `Builtin.isConcrete`), which is introduced in https://github.com/apple/swift/pull/26466.

This allows LLVM to generate efficient SIMD code in Debug builds, which may result in up to a 120x performance improvement for some operations such as addition. This also results in some performance improvements for Release builds since we no longer rely on the optimizer to auto-vectorize the loop code.

The public-facing API is only changed through underscored operations on protocols like `SIMDStorage` and introduced underscored types such as `_SIMDNever` and `_SIMDGenericNever<T>`.

### CODE REVIEW

**Comment by gottesmm** [comment]
Make sure you munge the stuff (i.e. get rid of IRGenPrepare/etc) before the review starts.

**Comment by nvzqz** [comment]
@gottesmm Can you elaborate why the IRGenPrepare stuff should go into IRGen directly?

**Comment by gottesmm** [comment]
Is there any reason that it /shouldn't/ be in IRGen?

File: include/swift/AST/Builtins.def [hunk 1]
```diff
@@ -51,30 +51,97 @@ BUILTIN_CAST_OR_BITCAST_OPERATION(SExtOrBitCast,  "sextOrBitCast",  "n")
...
+/// NOTE: Often times the builtin infrastructure wants to treat all
+/// binary operation builtins generic or not the same way. To ensure
+/// we support all use cases in the compiler, we do not declare the
+/// operations as part of this builtin since often times this macro is
+/// used to generic code. Instead, we stamp this out using the
+/// overloaded_static, polymorphic, and all suffixed operations.
 #ifndef BUILTIN_BINARY_OPERATION
-#define BUILTIN_BINARY_OPERATION(Id, Name, Attrs, Overload) \
-          BUILTIN(Id, Name, Attrs)
+#define BUILTIN_BINARY_OPERATION(Id, Name, Attrs) BUILTIN(Id, Name, Attrs)
```
Comment (stephentyrone) [inline include/swift/AST/Builtins.def]: Can't we just get rid of `BUILTIN_BINARY_OPERATION` and use `BUILTIN` instead, then?

Comment (gottesmm) [reply include/swift/AST/Builtins.def]: No. Sometimes you want to `#define` something just for Builtin Binary Operations and not all builtins.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**